witness/src/Dscp/Witness/Mempool/Logic.hs

```diff
@@ -0,0 +1,128 @@
+       , HasLens' ctx MempoolVar
+       , HasLens' ctx SD.SDActions
+       , HasLens' ctx SD.LoggingIO
+       , UnliftIO.MonadUnliftIO    m
+       )
+    => MempoolVar
+    -> GTxWitnessed
+    -> m ()
+addTxToMempool (Mempool pool conf, lock) tx = do
+    writeToMempool @ctx pool lock
+        $ Pool.processTxAndInsertToMempool conf tx
+
+takeTxsMempool
+    :: forall ctx m
+    .  ( MonadIO m
+       , MonadReader ctx m
+       , HasLens' ctx MempoolVar
+       , HasLens' ctx SD.SDActions
+       , HasLens' ctx SD.LoggingIO
+       , UnliftIO.MonadUnliftIO    m
+       )
+    => MempoolVar
+    -> m [GTxWitnessed]
+takeTxsMempool (Mempool pool _, lock) = do
+    txsWithUndos <- writeToMempool @ctx pool lock Pool.evictMempool
+    return (map fst txsWithUndos)
+
+isInMempool
+    :: forall ctx m
+    .  ( MonadIO m
+       , MonadReader ctx m
+       , HasLens' ctx MempoolVar
+       , HasLens' ctx SD.SDActions
+       , HasLens' ctx SD.LoggingIO
+       , UnliftIO.MonadUnliftIO    m
+       )
+    => MempoolVar
+    -> GTxWitnessed -> m Bool
+isInMempool (Mempool pool _, lock) tx = do
+    txsWithUndos <- readFromMempool @ctx pool lock (gets Pool.msTxs)
+    return (tx `elem` map fst txsWithUndos)
+
+type SDM =
+    SD.ERwComp
+        Conf.Exceptions
+        Conf.Ids
+        Conf.Values
+        (SD.IOCtx (AVLP.AVLChgAccum Conf.Ids Conf.Values))
+        (Pool.MempoolState
+            Conf.Ids
+            Conf.Values
+            (AVLP.AVLChgAccum Conf.Ids Conf.Values)
+            GTxWitnessed)
+
+readFromMempool
+    :: forall ctx m a
+    .  ( MonadIO m
+       , MonadReader ctx m
+       , HasLens' ctx MempoolVar
+       , HasLens' ctx SD.SDActions
+       , HasLens' ctx SD.LoggingIO
+       , UnliftIO.MonadUnliftIO    m
+       )
+    => Pool.Mempool Conf.Ids Conf.Values (AVLP.AVLChgAccum Conf.Ids Conf.Values) GTxWitnessed
+    -> Lock.SDLock
+    -> SDM a
+    -> m a
+readFromMempool pool lock action = do
+    actions <- view (lensOf @SD.SDActions)
+    logger  <- view (lensOf @SD.LoggingIO)
+    let dbActions = SD.dmaAccessActions $ SD.nsStateDBActions actions (AVLP.RememberForProof False)
+    Lock.readingSDLockOf lock $ do
+        SD.runRIO logger $ Pool.actionWithMempool pool dbActions action
+
+writeToMempool
+    :: forall ctx m a
+    .  ( MonadIO m
+       , MonadReader ctx m
+       , HasLens' ctx MempoolVar
+       , HasLens' ctx SD.SDActions
+       , HasLens' ctx SD.LoggingIO
+       , UnliftIO.MonadUnliftIO    m
+       )
+    => Pool.Mempool Conf.Ids Conf.Values (AVLP.AVLChgAccum Conf.Ids Conf.Values) GTxWitnessed
+    -> Lock.SDLock
+    -> SDM a
+    -> m a
+writeToMempool pool lock action = do
+    actions <- view (lensOf @SD.SDActions)
+    logger  <- view (lensOf @SD.LoggingIO)
+    let dbActions = SD.dmaAccessActions $ SD.nsStateDBActions actions (AVLP.RememberForProof False)
+    Lock.writingSDLockOf lock $ do
+        SD.runRIO logger $ Pool.actionWithMempool pool dbActions action
```

Comment:
Oh here we go again with this `runRIO logger`, couldn't they spend 5m to think of how bad this design is? Ugh.

Reply:
As far as I heard, they needed to implement loot-based logging in SD, and just shadowed most stuff from disciplina. And now we suffer ))